apps/aecore/src/aec_trees_proxy.erl

```diff
@@ -0,0 +1,592 @@
+            {error, Error}
+    after 5000 ->
+            {error, timeout}
+    end.
+
+-spec start_monitor(aec_trees:trees(), aetx_env:env(), context(), proplist()) ->
+          {ok, {pid(), reference()}}.
+-if(?OTP_RELEASE >= 23).
+start_monitor(Trees, Env, #{indexed := _} = Ctxt, Opts) ->
+    gen_server:start_monitor(?MODULE, {self(), Trees, Env, Ctxt}, []).
+-else.
+start_monitor(Trees, Env, #{indexed := _} = Ctxt, Opts) ->
+    _TStore = get_tstore(),
+    {ok, Pid} = gen_server:start(?MODULE, {self(), Trees, Env, Ctxt, Opts}, []),
+    MRef = monitor(process, Pid),
+    {ok, {Pid, MRef}}.
+-endif.
+
+%% ======================================================================
+%% Gen_server side
+%% ======================================================================
+
+-spec init({pid(), trees(), env(), context(), proplist()}) -> {ok, #st{}}.
+init({Parent, Trees, Env, #{ indexed := Indexed
+                           , deps    := Deps
+                           , claims  := Claims }, Opts}) ->
+    ArgF = fun(Type) ->
+                   aeu_mtrees:proxy_tree(?MODULE, #{ type => Type
+                                                   , pid  => self() })
+           end,
+    ProxyTrees = aec_trees:proxy_trees(ArgF),
+    DontVerify = proplists:get_value(dont_verify_signature, Opts, false),
+    Strict = proplists:get_value(strict, Opts, false),
+    Events = aetx_env:events(Env),
+    S0 = #st{ parent      = Parent
+            , clients     = #{}
+            , deps        = Deps
+            , claims      = Claims
+            , proxy_trees = ProxyTrees
+            , dontverify  = DontVerify
+            , strict      = Strict
+            , env         = Env
+            , trees       = Trees
+            , events      = Events },
+    {ok, start_workers(Indexed, ProxyTrees, Env, DontVerify, S0)}.
+
+handle_call({?MODULE, #req{} = Req}, {Pid, _} = From, #st{} = St) ->
+    handle_req(Req#req{from = From}, Pid, St).
+
+handle_cast(_, St) ->
+    {noreply, St}.
+
+handle_info({'DOWN', _MRef, process, Pid, Reason}, #st{ pids = Pids } = St) ->
+    case maps:find(Pid, Pids) of
+        {ok, Ix} ->
+            ?event({'DOWN', Pid, Ix, Reason}),
+            Pids1 = maps:remove(Pid, Pids),
+            ?event({pids1, Pids1}, St),
+            tx_down(Ix, Pid, Reason, St#st{pids = Pids1});
+        error ->
+            ?event({ignoring_DOWN, Pid}),
+            {noreply, St}
+    end;
+handle_info(_, St) ->
+    {noreply, St}.
+
+terminate(_Reason, _St) ->
+    ok.
+
+code_change(_FromVsn, St, _Extra) ->
+    {ok, St}.
+
+tx_down(_Ix, Pid, {'$caught', Type, What}, #st{strict = true} = St) ->
+    throw_error(Type, What, Pid, St);
+tx_down(Ix, _Pid, Reason, #st{ status = Status
+                             , events = Events0 } = St) ->
+    %% All valid txs must have at least some claims.
+    {IsValid, Events} =
+        case Reason of
+            {ok, NewEvents} ->
+                {true, Events0 ++ NewEvents};
```

Comment:
This is linear in the first list - this might become an issue when many events were emitted